### PR TITLE
Issue/refactor api notify send messages #342

### PR DESCRIPTION
**Pull request by JekaMas**
Refactor and little bit clean up Notify api:

 Created interface and package "notification" and extracted related code into it
 Set dependencies into constructor notificatotion.Manager
 Mocks for notificatotion.Manager and FCMClient
 Bacis unit tests for Notify and Send
 Notify func is now deprecated
 Notify users uses new API: message, Payload and a list of tokens

### CODE REVIEW

File: geth/common/notification.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+// Notification manages Push Notifications and send messages
+type Notification interface {
+	Notify(token string) string
+	Send() error
+}
+
+// Messaging manages send/notification messaging clients
+type Messaging interface {
+	NewFcmRegIdsMsg(list []string, body interface{}) *fcm.FcmClient
```
Comment (influx6) [inline geth/common/notification.go]: Can we rename this to `NewClient` or `NewFCMClient`? 

`NewFCMReqIdsMsg` does not really fit it's purpose and seem more of made up due to method's argument. I think naming methods with regards to purpose and return value is more idiomatic. :)

File: geth/notification/notification.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	"github.com/status-im/status-go/geth/common"
+	"github.com/status-im/status-go/geth/log"
+)
+
+// Manager of push notifications
+type Manager struct {
+	messaging common.Messaging
+}
+
+// New notifications manager
```
Comment (influx6) [inline geth/notification/notification.go]: `New returns new instance of notification manager`?

File: geth/common/notification.go (hunk 1 above)
Comment (JekaMas) [reply geth/common/notification.go]: This is copy of methods of "github.com/NaySoftware/go-fcm" to create interface on our side. So i need exact match of methods and cant change `NewFCMReqIdsMsg` interface.

File: geth/notification/notification.go (hunk 2 above)
Comment (JekaMas) [reply geth/notification/notification.go]: Yep.The usual call looks like 'notification.New()'. I think it tells by itself.

File: geth/common/notification.go (hunk 1 above)
Comment (influx6) [reply geth/common/notification.go]: Alright.

File: geth/common/notification.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package common
```
[the remark this answers is not in the document]
Comment (JekaMas) [reply geth/common/notification.go]: This is useful practice to prevent cyclic dependencies. But i agree that "common" package for interfaces looks strange and uninformative. We can do this refactoring in separete issue.

[the remark this answers is not in the document]
Comment (adambabik) [reply geth/common/notification.go]: > This is useful practice to prevent cyclic dependencies.

Right, but I doubt it's needed in most of the cases here.

> We can do this refactoring in separete issue.

Agree. It's better to be consistent for now.

File: geth/api/api.go [hunk 4]
```diff
@@ -195,32 +189,10 @@ func (api *StatusAPI) JailBaseJS(js string) {
...
-
-	c := fcm.NewFcmClient(serverKey)
-	c.NewFcmRegIdsMsg(ids, data)
-	c.SetNotificationPayload(&NP)
-
-	_, err := c.Send()
-	if err != nil {
-		log.Error("Notify failed:", err)
-	}
+	api.b.notification.Notify(token)
```
[the remark this answers is not in the document]
Comment (JekaMas) [reply geth/api/api.go]: We cant change interface 'Messaging'. It must be exactly the same as public methods in FCMClient for mocking. We cant mock a struct, but we can mock an interface.

Nice idea about provider i`m going to make that way!

About the splitting: notify makes notification, send makes send. This is just a single responsibility.

[the remark this answers is not in the document]
Comment (adambabik) [reply geth/api/api.go]: > We cant change interface 'Messaging'. It must be exactly the same as public methods in FCMClient for mocking.

We can't design an interface because we need a mock. We plan to support more than one provider and in such a case we would need to extend this interface or make a composition of many interfaces. 

Also, it's not clear what `Messaging` and `Notification` responsibilities are. Two interfaces that I can imagine are:
1. `Notifier` which can send notifications,
2. `NotificationProvider` which is a notification backend. In our case, `FCMClient` can be encapsulated in another `struct` which implements `NotificationProvider` interface.

> About the splitting: notify makes notification, send makes send. This is just a single responsibility.

`notify` does not imply "making notification" but rather both. If we want it to be divided then let's call them closer to: `CreateNotification(token, payload) Notification` and `Send(notification)`.

File: geth/notification/fcm_client_test.go [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	suite.Run(t, new(FCMClientTestSuite))
+}
+
+type FCMClientTestSuite struct {
+	t.BaseTestSuite
+}
+
+func (s *FCMClientTestSuite) TestNewFCMClient() {
+	fcmClient := NewFCMClient()
+	s.Require().NotNil(fcmClient)
```
[the remark this answers is not in the document]
Comment (adambabik) [reply geth/notification/fcm_client_test.go]: This comment shouldn't be here. It's regarding `s.Require().IsType(&fcm.FcmClient{}, fcmClient)`. I think we can assume that `NewFCMClient()` returns a proper type.

File: geth/api/api.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (JekaMas) [reply geth/api/api.go]: I think i got your point. I`m going try this in PR update.

File: geth/common/notification.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (JekaMas) [reply geth/common/notification.go]: I've added better interfaces, more tests and rewrite comments in the last commits. Could you review PR?

File: geth/api/backend.go [hunk 6]
```diff
@@ -34,12 +41,14 @@ func NewStatusBackend() *StatusBackend {
 	accountManager := account.NewManager(nodeManager)
 	txQueueManager := txqueue.NewManager(nodeManager, accountManager)
 	jailManager := jail.New(nodeManager)
+	notificationManager := notification.New(notification.NewFCMProvider(fcm.NewFcmClient(fcmServerKey)))
```
Comment (adambabik) [inline geth/api/backend.go]: Could it work like this:
```go
notification.NewFCMProvider(fcmServerKey)
```
so that we encapsulate client creation in `NewFCMProvider`?

File: geth/notification/fcm_provider.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
+package notification
+
+import (
+	"github.com/NaySoftware/go-fcm"
+	"github.com/status-im/status-go/geth/notification/message"
+	"github.com/status-im/status-go/geth/common"
+)
+
+// FCMProvider represents messaging provider for notifications.
```
Comment (adambabik) [inline geth/notification/fcm_provider.go]: Maybe we could be more descriptive here? For people who will open this code for the first time, FCM says nothing. Something like: `// FCMProvider represents Firebase Cloud Messaging provider for notifications.`

File: geth/notification/fcm_provider.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+type FCMProvider struct {
+	common.FirebaseClient
+}
+
+// NewFCMProvider Firebase Cloud Messaging client constructor.
+func NewFCMProvider(fcmClient common.FirebaseClient) *FCMProvider {
+	return &FCMProvider{fcmClient}
```
Comment (adambabik) [inline geth/notification/fcm_provider.go]: As I mentioned earlier, I guess `fcmClient` could be created in here.

File: geth/notification/fcm_provider.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+// SetMessage to send for given ids.
```
Comment (adambabik) [inline geth/notification/fcm_provider.go]: What are these ids?

File: geth/notification/fcm_provider.go [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+	p.NewFcmRegIdsMsg(ids, body)
+}
+
+// SetPayload sets payload message information.
+func (p *FCMProvider) SetPayload(payload *message.Payload) {
+	fcmPayload := p.toFCMPayload(payload)
+	p.SetNotificationPayload(fcmPayload)
+}
+
+// Send message.
```
Comment (adambabik) [inline geth/notification/fcm_provider.go]: `Send notification.`?

File: geth/notification/fcm_provider.go [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+import (
+	"github.com/NaySoftware/go-fcm"
+	"github.com/status-im/status-go/geth/notification/message"
+	"github.com/status-im/status-go/geth/common"
+)
+
+// FCMProvider represents messaging provider for notifications.
+type FCMProvider struct {
+	common.FirebaseClient
```
Comment (adambabik) [inline geth/notification/fcm_provider.go]: I don't understand why we need an interface here. Why not:
```go
type FCMProvider struct {
  client *fcm.Client
}
```

Is it to allow mocking? If so, check if it's possible to customize `fcm.FcmClient`'s server URL. If so, you can use an `httptest.Server`. This interface is not intuitive as it's used only in one place. 

If you decide to leave that interface, let's put it in this package instead of `common`.

File: geth/notification/notification.go [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+// New notifications manager.
+func New(messaging common.MessagingProvider) *Manager {
+	return &Manager{
+		messaging,
+	}
+}
+
+// Notify makes send message and notification.
+func (n *Manager) Notify(token string, msg *message.Message) (string, error) {
+	log.Debug("Notify", "token", token)
```
Comment (adambabik) [inline geth/notification/notification.go]: I think this kind of logs should be put in a caller.

File: geth/notification/notification.go [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+	n.messaging.SetMessage([]string{token}, msg.Body)
+	n.messaging.SetPayload(msg.Payload)
+
+	err := n.messaging.Send()
+	if err != nil {
+		log.Error("Notify failed:", err)
```
Comment (adambabik) [inline geth/notification/notification.go]: Alike, this log should be put in a caller.

File: geth/notification/notification.go [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+	}
+
+	return token, err
```
Comment (adambabik) [inline geth/notification/notification.go]: Token was passed as an argument. Why do we want to return it back unchanged?

File: geth/notification/notification_test.go [hunk 15]
```diff
@@ -0,0 +1,78 @@
+package notification
+
+import (
+	"testing"
+
+	"errors"
+	"github.com/golang/mock/gomock"
+	"github.com/status-im/status-go/geth/common"
+	"github.com/status-im/status-go/geth/notification/message"
+	t "github.com/status-im/status-go/geth/testing"
```
Comment (adambabik) [inline geth/notification/notification_test.go]: Confusing with `t *testing.T`. I would stay with `testing`. It's still pretty compact :)

File: geth/notification/notification_test.go [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+	s.messagingMock = common.NewMockMessagingProvider(s.messagingMockCtrl)
+}
+
+func (s *NotificationTestSuite) TearDownTest() {
+	s.messagingMockCtrl.Finish()
+}
+
+func (s *NotificationTestSuite) TestNewNotification() {
+	manager := New(nil)
+	s.Require().NotNil(manager)
```
Comment (adambabik) [inline geth/notification/notification_test.go]: You can write `s.NotNil(manager)`.

File: geth/notification/fcm_provider_test.go [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+func (s *FCMProviderTestSuite) TestSetMessage() {
+	ids := []string{"1"}
+	body := interface{}("body")
+
+	s.fcmClientMock.EXPECT().NewFcmRegIdsMsg(ids, body).Times(1)
+	fcmClient := NewFCMProvider(s.fcmClientMock)
+
+	fcmClient.SetMessage(ids, body)
```
Comment (adambabik) [inline geth/notification/fcm_provider_test.go]: Not sure why you decide to have a new line before this one.

Isn't this one easier to follow?
```go
ids := []string{"1"}
body := interface{}("body")

s.fcmClientMock.EXPECT().NewFcmRegIdsMsg(ids, body).Times(1)

fcmClient := NewFCMProvider(s.fcmClientMock)
fcmClient.SetMessage(ids, body)
```

File: geth/notification/notification_test.go (hunk 15 above)
Comment (JekaMas) [reply geth/notification/notification_test.go]: Little bit higher there is already import of standard 'testing' package.

**Comment by JekaMas** [comment]
The last summory:
i`ve removed provider interface. i think it does not make code better or more abstract, actually it pretty the same as Notifier interface;
FCM interface is now private and same level as implementation
ids -> tokens

File: geth/api/api.go [hunk 18]
```diff
@@ -195,29 +190,17 @@ func (api *StatusAPI) JailBaseJS(js string) {
...
-	}
-
-	c := fcm.NewFcmClient(serverKey)
-	c.NewFcmRegIdsMsg(ids, data)
-	c.SetNotificationPayload(&NP)
-
-	_, err := c.Send()
+	err := api.b.notification.Notify(msg, token)
 	if err != nil {
 		log.Error("Notify failed:", err)
```
Comment (tiabc) [inline geth/api/api.go]: Let's return an error to the caller?

File: geth/api/backend.go [hunk 19]
```diff
@@ -23,7 +29,7 @@ type StatusBackend struct {
 	accountManager common.AccountManager
 	txQueueManager common.TxQueueManager
 	jailManager    common.JailManager
-	// TODO(oskarth): notifer here
+	notification   common.Notifier
```
Comment (tiabc) [inline geth/api/backend.go]: I think having this field named `notifier` would ease understanding. Because it's not a single notification, it's a notifier which may send many notifications.

File: geth/notification/fcm/notifier.go [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+	firebaseClient
+}
+
+// NewNotifier Firebase Cloud Messaging client constructor.
+func NewNotifier(key string) *Notifier {
+	return &Notifier{fcm.NewFcmClient(key)}
+}
+
+// Notify preparation and send to the tokens list.
+func (p *Notifier) Notify(body interface{}, tokens ...string) error {
```
Comment (tiabc) [inline geth/notification/fcm/notifier.go]: Why is the receiver named `n`? Let's rename to `p`?

File: geth/notification/fcm/notifier.go [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+	p.setPayload(&notification.Payload{
+		Title: "Status - new message",
+		Body:  "ping",
+	})
+
+	p.setMessage(body, tokens...)
+	_, err := p.firebaseClient.Send()
```
Comment (tiabc) [inline geth/notification/fcm/notifier.go]: If you embed `firebaseClient` and use it transparently, let's use `p.Send()` for consistency?

File: geth/notification/fcm/notifier.go [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+// Notify preparation and send to the tokens list.
+func (p *Notifier) Notify(body interface{}, tokens ...string) error {
+	p.setPayload(&notification.Payload{
+		Title: "Status - new message",
+		Body:  "ping",
+	})
+
+	p.setMessage(body, tokens...)
```
Comment (tiabc) [inline geth/notification/fcm/notifier.go]: As I can see, this object is mutable and `Notify` signal is possible to be invoked in parallel. Can you, please, write a test which shows there's no race conditions? From the first glance, here will be some.

File: geth/api/api.go [hunk 23]
```diff
@@ -195,29 +190,17 @@ func (api *StatusAPI) JailBaseJS(js string) {
...
-	var NP fcm.NotificationPayload
-	NP.Title = "Status - new message"
-	NP.Body = "ping"
-
 	// TODO(oskarth): Experiment with this
-	data := map[string]string{
+	msg := map[string]string{
 		"msg": "Hello World1",
 		"sum": "Happy Day",
 	}
```
Comment (tiabc) [inline geth/api/api.go]: Just out of curiosity. How is this going to be unmarshalled on the receiving side and where is this information shown? I've seen a notification payload below. So what is this?
cc @oskarth

File: geth/api/api.go (hunk 18 above)
Comment (JekaMas) [reply geth/api/api.go]: We can return an error but this is an epi change. Is it OK for refactoring issue?

File: geth/api/backend.go (hunk 19 above)
Comment (JekaMas) [reply geth/api/backend.go]: Ok.

File: geth/notification/fcm/notifier.go (hunk 21 above)
Comment (JekaMas) [reply geth/notification/fcm/notifier.go]: Notifier has already Send() method so we cant just call `p.Send` to call embeded method `p.firebase.Send()`.

File: geth/notification/fcm/notifier.go (hunk 22 above)
Comment (JekaMas) [reply geth/notification/fcm/notifier.go]: ok

File: geth/api/api.go (hunk 18 above)
Comment (tiabc) [reply geth/api/api.go]: It should have been done at the very beginning.
@oskarth?

File: geth/api/api.go (hunk 23 above)
Comment (oskarth) [reply geth/api/api.go]: @tiabc @JekaMas There's a notification payload which is what is visible to end user, then there is a data payload (this) which isn't visible but can be used. Then there are some additional settings such as CLICK_ACTION for Android.

Btw, we can extend the interface to be more generic (expose setting for other client, notification, data payload, CLICK_ACTION and similar things) if you guys want. Perhaps this is better done in a separate commit though. If you want to go down that route, we can make the necessary changes in status-react at the same time just so we have that capability, while keeping the same behavior as we have now.

Comment (oskarth) [reply geth/api/api.go]: See https://github.com/status-im/status-react/issues/1883 and https://github.com/status-im/status-react/issues/1907 for some context from status-react side

File: geth/api/api.go (hunk 18 above)
Comment (adambabik) [reply geth/api/api.go]: I didn't notice that. I guess there is no point in returning the same value (`token`) which was passed an argument.

Comment (oskarth) [reply geth/api/api.go]: Feel free to change this to whatever makes sense, it was purely there as I spiked things out for debugging.

File: geth/api/api.go (hunk 23 above)
Comment (tiabc) [reply geth/api/api.go]: Thanks for the explanation and some related background, @oskarth!
What did you mean about our desire to make it more generic? If there's a requirements for that, we'd be happy to make it more general, just file an issue for that :)

File: geth/api/api.go (hunk 18 above)
Comment (JekaMas) [reply geth/api/api.go]: Ok, i`m going to return {Status: bool, Error: string}.

File: geth/api/api.go (hunk 23 above)
Comment (JekaMas) [reply geth/api/api.go]: Actually i`ve make payload on our side: https://github.com/status-im/status-go/blob/028206653571379c3f7d6e2ce115c6217b02f80d/geth/notification/payload.go

Comment (oskarth) [reply geth/api/api.go]: @tiabc No requirement right now.

**Comment by JekaMas** [comment]
@tiabc i did PR update few days ago, please review it.

File: cmd/statusd/library.go [hunk 24]
```diff
@@ -386,6 +386,25 @@ func makeJSONResponse(err error) *C.char {
...
+		res = false
+		errString = err.Error()
+	}
+
+	out := common.NotifyResult{
+		Status: res,
+		Error:  errString,
+	}
+
+	outBytes, err := json.Marshal(&out)
```
Comment (tiabc) [inline cmd/statusd/library.go]: Why a reference to `out` here?

File: geth/api/backend.go [hunk 25]
```diff
@@ -29,7 +29,7 @@ type StatusBackend struct {
 	accountManager common.AccountManager
 	txQueueManager common.TxQueueManager
 	jailManager    common.JailManager
-	notification   common.Notifier
+	notifier       common.NotifierConstructor
```
Comment (tiabc) [inline geth/api/backend.go]: Why changing it here?

Comment (JekaMas) [reply geth/api/backend.go]: I've considered 'no state == no race' and changed constructor to higher order function to make new Notify object every time. Notify object is small and we can make it fast and safe, with no races risk.

File: cmd/statusd/library.go (hunk 24 above)
Comment (JekaMas) [reply cmd/statusd/library.go]: Hmm... i think it's copy-paste from this file.

File: geth/notification/fcm/notifier.go [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+	"github.com/status-im/status-go/geth/notification"
+)
+
+// Notifier represents messaging provider for notifications.
+type Notifier struct {
+	client firebaseClient
+}
+
+// NewNotifier Firebase Cloud Messaging client constructor.
+func NewNotifier(key string) func() common.Notifier {
```
Comment (tiabc) [inline geth/notification/fcm/notifier.go]: I suppose the return type in the function signature should be `common.NotifierConstructor`? Technically they're almost the same but this option is better for understanding.

Also, I'm still trying to come up with proper naming. Now that you've made it a constructor (which is definitely a good idea!), it's not `Notifier` any more. It's `Notification` and should have a single `Send` method instead of `Notify`.
Here https://github.com/status-im/status-go/pull/390/files#diff-ebd449640936a8f327a6a4a1c3274d7cR32 I'd also named it `newNotification` so that the usage is `api.b.newNotification()`. By the way, can't we just create a separate constructor for `Notification` and just call it before sending a notification? It'll be a way simpler approach.

What do you think of all that?

File: cmd/statusd/library.go (hunk 24 above)
Comment (tiabc) [reply cmd/statusd/library.go]: Let's fix it?

File: cmd/statusd/library.go [hunk 27]
```diff
@@ -386,6 +386,25 @@ func makeJSONResponse(err error) *C.char {
...
+
+	res := true
+	errString := ""
+	if err != nil {
+		res = false
+		errString = err.Error()
+	}
+
+	out := common.NotifyResult{
+		Status: res,
```
Comment (adambabik) [inline cmd/statusd/library.go]: You can write it as `Status: (err == nil)` to avoid extra variables.

File: geth/common/types.go [hunk 28]
```diff
@@ -395,6 +395,12 @@ type TestConfig struct {
 	}
 }
 
+// NotifyResult is a JSON returned from notify message
+type NotifyResult struct {
+	Status bool   `json:"status"`
+	Error  string `json:"error"`
```
Comment (adambabik) [inline geth/common/types.go]: Add `omitempty`? I guess there is no point it passing this field with an empty string. 

It's still something that we need to sit and talk about with ClojureTeam as it's pretty inconsistent.

Comment (tiabc) [reply geth/common/types.go]: You can start: https://docs.google.com/document/d/1BBit9U2_-_B12NczUiOGzrGMU8z1vQeAAuqE2HVIYKc :)

Comment (adambabik) [reply geth/common/types.go]: What do you think about making this doc more general and actually specify input and output formats? Specifying error codes won't change much as we still won't have a complete picture what kind of data we exchange.

Comment (tiabc) [reply geth/common/types.go]: I've just started it when we had a similar question so that somebody filled it :) I'll make an issue for that shortly.

Comment (adambabik) [reply geth/common/types.go]: I see :) Ok, let's make an issue to track progress, keep all related issues in one place and define acceptance criteria. It will be super useful when we decide to switch to gomobile.

Comment (tiabc) [reply geth/common/types.go]: #408

File: cmd/statusd/library.go (hunk 24 above)
Comment (JekaMas) [reply cmd/statusd/library.go]: Ok, done.

File: geth/notification/fcm/notifier.go (hunk 26 above)
Comment (JekaMas) [reply geth/notification/fcm/notifier.go]: Done.

File: geth/common/types.go (hunk 28 above)
Comment (JekaMas) [reply geth/common/types.go]: 'omitempty' for error added.

File: cmd/statusd/library.go [hunk 29]
```diff
@@ -21,7 +21,7 @@ func GenerateConfig(datadir *C.char, networkID C.int, devMode C.int) *C.char {
 		return makeJSONResponse(err)
 	}
 
-	outBytes, err := json.Marshal(&config)
+	outBytes, err := json.Marshal(config)
```
Comment (tiabc) [inline cmd/statusd/library.go]: Just notifying that generally we must not make unrelated changes because they may impede merging sometimes. This change won't interfere with anything else so it's fine.

File: geth/api/backend.go [hunk 30]
```diff
@@ -34,12 +40,14 @@ func NewStatusBackend() *StatusBackend {
 	accountManager := account.NewManager(nodeManager)
 	txQueueManager := txqueue.NewManager(nodeManager, accountManager)
 	jailManager := jail.New(nodeManager)
+	notificationManager := fcm.NewNotifier(fcmServerKey)
```
Comment (tiabc) [inline geth/api/backend.go]: Wouldn't a separate method be simpler than making a function returning function accessing a previously passed argument as in a closure?

File: geth/notification/fcm/notifier.go [hunk 31]
```diff
@@ -0,0 +1,55 @@
+package fcm
+
+import (
+	"github.com/NaySoftware/go-fcm"
+	"github.com/status-im/status-go/geth/common"
+	"github.com/status-im/status-go/geth/notification"
+)
+
+// Notifier represents messaging provider for notifications.
+type Notifier struct {
```
Comment (tiabc) [inline geth/notification/fcm/notifier.go]: As I wrote before, it's not `Notifier` any more, it's `Notification`, right?

File: geth/api/backend.go (hunk 30 above)
Comment (JekaMas) [reply geth/api/backend.go]: I dont think so, higher order function looks cleaner and more safe, isnt it?

File: geth/notification/fcm/notifier.go (hunk 31 above)
Comment (JekaMas) [reply geth/notification/fcm/notifier.go]: Ok, i just think that we want only rename constructor.

File: geth/api/backend.go (hunk 30 above)
Comment (tiabc) [reply geth/api/backend.go]: Come on :)

File: geth/notification/fcm/notifier.go (hunk 31 above)
Comment (tiabc) [reply geth/notification/fcm/notifier.go]: We want names to be compliant with entities essence :)

File: geth/api/backend.go (hunk 30 above)
Comment (tiabc) [reply geth/api/backend.go]: And what about this? Obviously, higher-order functions only adds complexity, I don't see why we need it here instead of just using a constructor, can you explain, please?

Comment (JekaMas) [reply geth/api/backend.go]: With higher order function:
```
//Declaration
notifier := fcm.NewNotifier(fcmServerKey)

// Usage
err := api.b.notifier().Send(msg, token)
```

Without:
```
diff --git a/geth/api/api.go b/geth/api/api.go
index 170faf40..d519cda7 100644
--- a/geth/api/api.go
+++ b/geth/api/api.go
@@ -7,6 +7,7 @@ import (
 	gethcommon "github.com/ethereum/go-ethereum/common"
 	"github.com/status-im/status-go/geth/common"
 	"github.com/status-im/status-go/geth/log"
+	"github.com/status-im/status-go/geth/notification/fcm"
 	"github.com/status-im/status-go/geth/params"
 )
 
@@ -200,7 +201,8 @@ func (api *StatusAPI) Notify(token string) error {
 		"sum": "Happy Day",
 	}
 
-	err := api.b.newNotification().Send(msg, token)
+	notification := fcm.NewNotification(fcmServerKey)
+	err := notification.Send(msg, token)
 	if err != nil {
 		log.Error("Notify failed:", err)
 	}
diff --git a/geth/api/backend.go b/geth/api/backend.go
index 302e8b50..b0d318fa 100644
--- a/geth/api/backend.go
+++ b/geth/api/backend.go
@@ -24,12 +24,12 @@ const (
 // StatusBackend implements Status.im service
 type StatusBackend struct {
 	sync.Mutex
-	newNotification common.NotificationConstructor
+	notification   common.Notification
 }
 
 // NewStatusBackend create a new NewStatusBackend instance
@@ -40,14 +40,14 @@ func NewStatusBackend() *StatusBackend {
-	notificationManager := fcm.NewNotification(fcmServerKey)
+	notification := fcm.NewNotification(fcmServerKey)
 
 	return &StatusBackend{
-		newNotification: notificationManager,
+		notification:   notification,
 	}
 }
 
diff --git a/geth/common/notification.go b/geth/common/notification.go
index 34ba674c..e58c142a 100644
--- a/geth/common/notification.go
+++ b/geth/common/notification.go
@@ -1,9 +1,6 @@
 package common
 
-// Notifier manages Push Notifications.
-type Notifier interface {
+// Notification manages Push Notifications.
+type Notification interface {
 	Send(body interface{}, tokens ...string) error
 }
-
-// NotificationConstructor returns constructor of configured instance Notifier interface.
-type NotificationConstructor func() Notifier
diff --git a/geth/notification/fcm/notification.go b/geth/notification/fcm/notification.go
index 62101ece..2c8e9bad 100644
--- a/geth/notification/fcm/notification.go
+++ b/geth/notification/fcm/notification.go
@@ -2,7 +2,6 @@ package fcm
 
 import (
 	"github.com/NaySoftware/go-fcm"
-	"github.com/status-im/status-go/geth/common"
 	"github.com/status-im/status-go/geth/notification"
 )
 
@@ -12,10 +11,8 @@ type Notification struct {
 }
 
 // NewNotification Firebase Cloud Messaging client constructor.
-func NewNotification(key string) common.NotificationConstructor {
-	return func() common.Notifier {
-		return &Notification{fcm.NewFcmClient(key)}
-	}
+func NewNotification(key string) *Notification {
+	return &Notification{fcm.NewFcmClient(key)}
 }
```

Now StatusApi should know about `github.com/status-im/status-go/geth/notification/fcm` and `fcmServerKey` and we can't easy test StatusAPI.Notify method.

**Comment by influx6** [comment]
Please fix conflicts. I dont have anything new to add on this, so good work!

File: geth/api/backend.go (hunk 30 above)
Comment (tiabc) [reply geth/api/backend.go]: I got your point. Pretty neat, actually!

File: geth/api/backend.go [hunk 32]
```diff
@@ -10,20 +10,26 @@ import (
...
-	accountManager common.AccountManager
-	txQueueManager common.TxQueueManager
-	jailManager    common.JailManager
-	// TODO(oskarth): notifer here
+	nodeReady       chan struct{} // channel to wait for when node is fully ready
+	nodeManager     common.NodeManager
+	accountManager  common.AccountManager
+	txQueueManager  common.TxQueueManager
+	jailManager     common.JailManager
+	newNotification common.NotificationConstructor
```
Comment (adambabik) [inline geth/api/backend.go]: I still have no idea why we couldn't stay with `common.NotificationManager`. We are introducing a new keyword `Constructor` and it's gonna be hard to explain it to newcomers. 

A call:
```go
err := api.b.newNotification().Send(msg, token)
```
would be
```go
err := api.b.notificationManager.GetClient().Send(msg, token)
```

File: geth/notification/fcm/notification.go [hunk 33]
```diff
@@ -0,0 +1,55 @@
+package fcm
+
+import (
+	"github.com/NaySoftware/go-fcm"
+	"github.com/status-im/status-go/geth/common"
+	"github.com/status-im/status-go/geth/notification"
+)
+
+// Notification represents messaging provider for notifications.
+type Notification struct {
```
Comment (adambabik) [inline geth/notification/fcm/notification.go]: `Client` would be easier to reason about. Seeing `Notification` type, I imagine a struct with some fields that is an actual notification, not something that sends notifications.

File: geth/api/backend.go (hunk 32 above)
Comment (JekaMas) [reply geth/api/backend.go]: We need to get new one notification for each request or it'll be data race. This is the first reason. The second one, actually i think
`err := api.b.newNotification().Send(msg, token)`
is much cleaner: it doesnt expose internal interfaces or parts like `.GetClient`.

**Comment by JekaMas** [comment]
@tiabc All conflicts has been resolved.

**Comment by tiabc** [comment]
Merging as it's been hanging for a while. We'll sort out any arising problems later, they're sort of minor.
